raster-foundry/raster-foundry
# Improve scene list query so it doesn't crash the database

## Overview

This PR listing scenes with their related objects. Previously, we used a lot of subqueries and common table
expressions, and it made the database very upset (like, many gb of temp files upset) and killed all of our freeable memory and CPU credits. This change does a few things to address that and some other ergonomics issues:

- no more nested queries -- instead of having giant sql queries, we use functional programming to combine the results of much simpler sql expressions
- split out `ImageWithRelatedDao` (new) and `SceneWithRelatedDao` (not new) to separate files to make parsing their file contents easier
- get rid of empty fragments when trying to construct the `ORDER BY` piece of paginated requests, since intercalating commas in `List(Fragment.empty, Fragment.empty)` pretty reasonably gives you the string `" , "`, which caused syntax errors if you typed the ordering field wrong

### Checklist

- ~Swagger specification updated, if necessary~
- ~Symlinks from new migrations present or corrected for any new migrations~

### Notes

This isn't a magic bullet -- some performance is still quite bad, for example, anything that requires a sort on the whole table. It may be reasonable to think of that as a fact of life. Sorts on result sets
of 50k scenes were reasonably quick, or at least quick enough that I didn't feel impatient. Most
result sets limited to geographic areas with cloud cover filters and with datasources selected will
be considerably smaller than that.

## Testing Instructions

  * create an ssh tunnel with the database's port forwarded (see Hector's threads in slack for instructions on how)
  * create a new transactor with the staging connection information (see [Doobie docs](https://tpolecat.github.io/doobie/docs/03-Connecting.html) for how if you're not sure)
 * start an sbt shell outside the vm, or wherever you have the port forwarded to and get set up to run some queries:

```scala
import com.lonelyplanet.akka.http.extensions.{PageRequest, Order}
// bogus parameter to prove that mistyped parameters don't ruin everything in a totally opaque way
// maybe they should though -- I think they should but didn't want to change too much at once
val pr = PageRequest(100, 20, Map("acquisitionDateTime" -> Order.Desc, "foo" -> Order.Asc))
val csq = CombinedSceneQueryParams(sceneParams=SceneQueryParameters(maxCloudCover=Some(0.3f)))
val xa2 = Transactor.fromDriverManager[IO](...)
```

 * get a user and attempt to list some scenes:

```scala
val u = UserDao.getUserById("default").transact(xa2).unsafeRunSync.get
SceneWithRelatedDao.listScenes(pr, csq, u).transact(xa2).unsafeRunSync
```

 * it should succeed in a reasonable amount of time

Closes #3084
